### PR TITLE
Bug 1873140: fix Alert State filter for dev perspective Alerts

### PR DESCRIPTION
**Pull request by vikram-raj**
Fix - https://issues.redhat.com/browse/ODC-4424

**Screenshots**
Kapture 2020-09-03 at 20 17 13

**Test coverage**
Screenshot 2020-08-27 at 6 45 29 PM

### CODE REVIEW

**Comment by kyoto** [comment]
I realized that we should be making a similar change to the admin perspective too. I will take a look into that and consider if the two perspectives can share code for this fix.

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx
Comment (kyoto) [inline frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:181]: Why only take the state of the first alert? Don't we need to iterate through all alerts looking for any alert that has a matching state?

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:170]: Maybe its just for myself to understand this.. If none filter is selected, you use the o (entry in filteredObjects) as filter value. Isn't it easier and cleaner to wrap this in an if, in the case we want keep 'all' filteredObjects? WDYT? :thinking: 

```jsx
    if (_.isFunction(filter)) {
      if (name === 'alert-state') {
        if (_.isEmpty(value.selected)) {
          filteredObjects = _.filter(filteredObjects, (o) =>
            _.filter(o.alerts, (a) => filter(value, a)).length > 0 ||
              (value.selected.has('inactive') && _.isEmpty(o.alerts)),
          );
        }
      } else {
        filteredObjects = _.filter(filteredObjects, (o) => filter(value, o));
      }
    }
```

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx
Comment (jerolimov) [reply frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:181]: Or if not, you could just write rule.alerts[0]?.state, but I think @kyoto question here is valid

**Comment by jerolimov** [comment]
Could verify the fix on a cluster setup. Thanks @vikram-raj for providing some help and yaml files here. Will get a lgtm when you response to the open questions :smirk:

**Comment by kyoto** [comment]
Opened a similar change for the admin perspective: https://github.com/openshift/console/pull/6460. The solutions will need to be different because the dev perspective combines rule rows and alert rows in the same table, but the admin perspective does not.

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:181]: I am mapping through all the alerts of the rule and it returns an array of a single element that is why I am taking the first element.

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:170]: Agree, Thanks for the suggestion.

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx
Comment (jerolimov) [reply frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:183]: But why mapping first over all elements and converting alerts to states, and then picking the first state. You could also pick the first alert state directly, or?

```js
  return rule.alerts[0].state || 'inactive';
```

I did not understand if we could reduce the alerts in any way instead of picking just the first one. /cc @kyoto

**Comment by vikram-raj** [comment]
/retitle Bug 1873140: fix Alert State filter for dev perspective Alerts

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx
Comment (kyoto) [reply frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:183]: It looks to me like this won't work when a rule has multiple active alerts with different states. For example, `_.map(rule.alerts, (a) => a.state)` could return `['firing', 'pending']`, and then this would ignore the `pending` alert.

File: frontend/packages/dev-console/src/components/monitoring/alerts/__tests__/monitoring-alerts-utils.spec.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/monitoring/alerts/__tests__/monitoring-alerts-utils.spec.tsx:19]: Can you keep the tests for both rules here?

```jsx
    alertRules[0].alerts[0].state = AlertStates.Inactive; // or what ever
    alertRules[1].alerts[0].state = AlertStates.Pending;
```

File: frontend/packages/dev-console/src/components/monitoring/alerts/__tests__/monitoring-alerts-utils.spec.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/monitoring/alerts/__tests__/monitoring-alerts-utils.spec.tsx:21]: Same here? Why this changed? :thinking:

File: frontend/public/components/monitoring/alerting.tsx
Comment (jerolimov) [inline frontend/public/components/monitoring/alerting.tsx:1310]: Please use lowercase a for variables (which are not a react component)

```suggestion
export const alertStateFilter: RowFilter = {
```

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:155]:
```suggestion
export const getFilteredRows = (_filters, rowFilters, rules: Rule[]) => {
```

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:161]:
```suggestion
  let filteredRules = rules;
```

and so on..

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:170]:
```suggestion
              _.some(o.alerts, (a) => filter(value, a)) ||
```

wdyt?

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx
Comment (jerolimov) [reply frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx]: Great final solution with `inMatch`. Nice work @vikram-raj and @kyoto

File: frontend/packages/dev-console/src/components/monitoring/alerts/__tests__/monitoring-alerts-utils.spec.tsx
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/monitoring/alerts/__tests__/monitoring-alerts-utils.spec.tsx]: Added the tests for other states as well.

File: frontend/packages/dev-console/src/components/monitoring/alerts/__tests__/monitoring-alerts-utils.spec.tsx
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/monitoring/alerts/__tests__/monitoring-alerts-utils.spec.tsx:21]: handle the test for rows 0 as well.

**Review by jerolimov — APPROVED** [review]
Tested it locally and on an cluster (4.6.0-0.ci.test-2020-09-07-093240-ci-ln-fl8bwh2), works as expected. :+1: 

/lgtm

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx]: This function is not required. So, removed it.

**Review by jerolimov — APPROVED** [review]
Tested it again on 4.6.0-0.ci.test-2020-09-07-115219-ci-ln-qgpzqit

Thanks @vikram-raj for the support here :)

/lgtm

**Pull request merged**
